logstash-core/lib/logstash/agent.rb

```diff
@@ -11,78 +12,167 @@
-    return 1
-  ensure
-    Stud::untrap("INT", sigint_id) unless sigint_id.nil?
-    Stud::untrap("TERM", sigterm_id) unless sigterm_id.nil?
-  end # def execute
+    if @auto_reload
+      Stud.interval(@reload_interval) { reload_state! }
+    else
+      while !Stud.stop?
+        if clean_state? || running_pipelines?
+          sleep 0.5
+        else
+          break
+        end
+      end
+    end
+  end
 
-  def add_pipeline(pipeline_id, config_str, settings = {})
-    @pipelines[pipeline_id] = LogStash::Pipeline.new(config_str, settings.merge(:pipeline_id => pipeline_id))
+  # register_pipeline - adds a pipeline to the agent's state
+  # @param pipeline_id [String] pipeline string identifier
+  # @param settings [Hash] settings that will be passed when creating the pipeline.
+  #   keys should be symbols such as :pipeline_workers and :pipeline_batch_delay
+  def register_pipeline(pipeline_id, settings)
+    pipeline = create_pipeline(settings.merge(:pipeline_id => pipeline_id))
+    return unless pipeline.is_a?(LogStash::Pipeline)
+    @pipelines[pipeline_id] = pipeline
   end
 
-  def node_uuid
-    @node_uuid ||= SecureRandom.uuid
+  def reload_state!
+    @upgrade_mutex.synchronize do
+      @pipelines.each do |pipeline_id, _|
+        begin
+          reload_pipeline!(pipeline_id)
+        rescue => e
+          @logger.error I18n.t("oops", :error => e, :backtrace => e.backtrace)
+        end
+      end
+    end
+  end
+
+  def shutdown
+    shutdown_pipelines
   end
 
   private
-  # Emit a warning message.
-  def warn(message)
-    # For now, all warnings are fatal.
-    raise LogStash::ConfigurationError, message
-  end # def warn
+  def node_uuid
+    @node_uuid ||= SecureRandom.uuid
+  end
 
-  # Emit a failure message and abort.
-  def fail(message)
-    raise LogStash::ConfigurationError, message
-  end # def fail
+  def create_pipeline(settings)
+    begin
+      config = fetch_config(settings)
+    rescue => e
+      @logger.error("failed to fetch pipeline configuration", :message => e.message)
+      return
+    end
 
-  def shutdown_pipelines
-    @pipelines.each do |_, pipeline|
-      pipeline.shutdown do
-        ::LogStash::ShutdownWatcher.start(pipeline)
-      end
+    begin
+      LogStash::Pipeline.new(config, settings)
+    rescue => e
+      @logger.error("fetched an invalid config", :config => config, :reason => e.message)
+      return
     end
   end
 
-  def trap_sigterm
-    Stud::trap("TERM") do
-      @logger.warn(I18n.t("logstash.agent.sigterm"))
-      shutdown_pipelines
+  def fetch_config(settings)
+    @config_loader.format_config(settings[:config_path], settings[:config_string])
+  end
+
+  # since this method modifies the @pipelines hash it is
+  # wrapped in @upgrade_mutex in the parent call `reload_state!`
+  def reload_pipeline!(id)
+    old_pipeline = @pipelines[id]
+    new_pipeline = create_pipeline(old_pipeline.original_settings)
```

Comment:
I'm glad you noticed this. I struggled a bit to find a place the pipeline settings needed for regenerating the pipeline. 
There are 2 sets of settings for the pipeline: the fetching instructions and the parametrization.
The pipeline instance must own the parametrization settings for its runtime, but not the fetching settings (like the cli args `config.path` and `config.string`) since the constructor receives the final "pipeline description string".

So I first tried with a hash that contained pipeline_id, fetch_settings and pipeline instance:

```ruby
@pipeline = {
  "pipelineid1" => {
     :pipeline => LogStash::Pipeline,
     :settings => { "config.path" => "/tmp/config" , ... },
  "pipelineid2" => {
      ...
  }
}
```

But creating a new Pipeline still required getting the runtime settings from the old pipeline and merging with the fetch settings so I opted to put them all "inside".
Also this hash made it awkward to interact with the pipeline object (e.g. `@pipelines["pipelineid1"].pipeline.shutdown`)

I therefore chose to have just:

```ruby
@pipeline = {
  "pipelineid1" => LogStash::Pipeline,
  "pipelineid2" => LogStash::Pipeline
}
```

An alternative here is to have a...*shudder*..pipeline factory class, that harbors the settings and produces pipelines, given the factory's settings.